File: miniprogram/game/GameBoard.js
export const SIZE = 15
export const EMPTY = 0
export const BLACK = 1
export const WHITE = 2

function createBoard() {
  const board = []
  for (let i = 0; i < SIZE; i++) {
    board[i] = []
    for (let j = 0; j < SIZE; j++) {
      board[i][j] = EMPTY
    }
  }
  return board
}

export class GameBoard {
  constructor() {
    this.board = []
    this.currentPlayer = BLACK
    this.gameOver = false
    this.winner = EMPTY
    this.winLine = null
    this.moveHistory = []
    this.reset()
  }

  reset() {
    this.board = createBoard()
    this.currentPlayer = BLACK
    this.gameOver = false
    this.winner = EMPTY
    this.winLine = null
    this.moveHistory = []
  }

  placeStone(row, col) {
    if (this.gameOver || row < 0 || row >= SIZE || col < 0 || col >= SIZE) return false
    if (this.board[row][col] !== EMPTY) return false
    this.board[row][col] = this.currentPlayer
    this.moveHistory.push({ row: row, col: col, player: this.currentPlayer })
    if (this.checkWin(row, col)) {
      this.winner = this.currentPlayer
      this.gameOver = true
    } else {
      this.currentPlayer = this.currentPlayer === BLACK ? WHITE : BLACK
    }
    return true
  }

  checkWin(row, col) {
    var dirs = [[0, 1], [1, 0], [1, 1], [1, -1]]
    var player = this.board[row][col]
    for (var d = 0; d < dirs.length; d++) {
      var dr = dirs[d][0], dc = dirs[d][1]
      var count = 1
      var sr = row, sc = col, er = row, ec = col
      for (var i = 1; i < 5; i++) {
        var r = row + dr * i, c = col + dc * i
        if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && this.board[r][c] === player) {
          count++; er = r; ec = c
        } else break
      }
      for (var i = 1; i < 5; i++) {
        var r = row - dr * i, c = col - dc * i
        if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && this.board[r][c] === player) {
          count++; sr = r; sc = c
        } else break
      }
      if (count >= 5) {
        this.winLine = { sr: sr, sc: sc, er: er, ec: ec }
        return true
      }
    }
    return false
  }

  undoLastMove() {
    if (this.moveHistory.length === 0) return false
    var last = this.moveHistory.pop()
    this.board[last.row][last.col] = EMPTY
    this.currentPlayer = last.player
    this.gameOver = false
    this.winner = EMPTY
    this.winLine = null
    return true
  }

  getCell(row, col) {
    return this.board[row][col]
  }

  getBoardCopy() {
    var copy = []
    for (var i = 0; i < SIZE; i++) {
      copy[i] = []
      for (var j = 0; j < SIZE; j++) {
        copy[i][j] = this.board[i][j]
      }
    }
    return copy
  }
}

File: lib/utils/game_board.dart
/// 五子棋棋盘逻辑
class GameBoard {
  static const int size = 15;
  static const int empty = 0;
  static const int black = 1;
  static const int white = 2;

  late List<List<int>> board;
  int currentPlayer = black;
  bool gameOver = false;
  int winner = empty;
  List<List<int>>? winLine;
  final List<List<int>> moveHistory = [];

  GameBoard() {
    board = List.generate(size, (_) => List.filled(size, empty));
    reset();
  }

  void reset() {
    for (var row in board) {
      row.fillRange(0, size, empty);
    }
    currentPlayer = black;
    gameOver = false;
    winner = empty;
    winLine = null;
    moveHistory.clear();
  }

  /// 在指定位置落子，返回是否成功
  bool placeStone(int row, int col) {
    if (gameOver || row < 0 || row >= size || col < 0 || col >= size) {
      return false;
    }
    if (board[row][col] != empty) {
      return false;
    }
    board[row][col] = currentPlayer;
    moveHistory.add([row, col, currentPlayer]);

    if (checkWin(row, col)) {
      winner = currentPlayer;
      gameOver = true;
    } else {
      currentPlayer = (currentPlayer == black) ? white : black;
    }
    return true;
  }

  bool checkWin(int row, int col) {
    const directions = [[0, 1], [1, 0], [1, 1], [1, -1]];
    final player = board[row][col];

    for (final dir in directions) {
      final dr = dir[0], dc = dir[1];
      int count = 1;
      int startRow = row, startCol = col;
      int endRow = row, endCol = col;

      for (int i = 1; i < 5; i++) {
        final r = row + dr * i, c = col + dc * i;
        if (r >= 0 && r < size && c >= 0 && c < size && board[r][c] == player) {
          count++;
          endRow = r;
          endCol = c;
        } else {
          break;
        }
      }
      for (int i = 1; i < 5; i++) {
        final r = row - dr * i, c = col - dc * i;
        if (r >= 0 && r < size && c >= 0 && c < size && board[r][c] == player) {
          count++;
          startRow = r;
          startCol = c;
        } else {
          break;
        }
      }

      if (count >= 5) {
        winLine = [[startRow, startCol], [endRow, endCol]];
        return true;
      }
    }
    return false;
  }

  int getCell(int row, int col) => board[row][col];

  /// 返回棋盘深拷贝，供AI线程安全使用
  List<List<int>> getBoardCopy() {
    return board.map((row) => List<int>.from(row)).toList();
  }

  /// 悔一步棋
  bool undoLastMove() {
    if (moveHistory.isEmpty) return false;
    final last = moveHistory.removeLast();
    board[last[0]][last[1]] = empty;
    currentPlayer = last[2];
    gameOver = false;
    winner = empty;
    winLine = null;
    return true;
  }
}
